src/js/BookReader/Mode1Up.js

```diff
@@ -12,144 +12,219 @@ export class Mode1Up {
-      if (prev) indicesToDisplay.unshift(firstIndexToDraw = prev.index);
-
-      const lastIndexToDraw = indicesToDisplay[indicesToDisplay.length - 1];
-      const next = book.getPage(lastIndexToDraw).findNext({ combineConsecutiveUnviewables: true });
-      if (next) indicesToDisplay.push(next.index);
+      yield { page, top: leafTop, bottom: leafBottom };
+      leafTop += height + this.physicalInchesToDisplayPixels(leafSpacing, reduce);
+      leafBottom += this.physicalInchesToDisplayPixels(leafSpacing, reduce);
     }
+  }
 
-    const BRpageViewEl = this.br.refs.$brPageViewEl.get(0);
-    leafTop = 0;
+  /**
+   * @param {{ top: number; bottom: number; }} bound1
+   * @param {{ top: number; bottom: number; }} bound2
+   */
+  static boundsIntersect(bound1, bound2) {
+    return (bound1.bottom >= bound2.top) && (bound1.top <= bound2.bottom);
+  }
 
-    for (const page of book.pagesIterator({ end: firstIndexToDraw, combineConsecutiveUnviewables: true })) {
-      leafTop += Math.floor(page.height / this.br.reduce) + 10;
+  /**
+   * Find the pages that intersect the current viewport, including 1 before/after
+   **/
+  * findIntersectingPages() {
+    // Rectangle of interest
+    const height = this.$scrollContainer.height();
+    const scrollTop = this.$scrollContainer.scrollTop();
+    const scrollBottom = scrollTop + height;
+    const scrollRegion = { top: scrollTop, bottom: scrollBottom };
+
+    let prev = null;
+    for (const {page, top, bottom} of this.pagesWithBounds()) {
+      const intersects = Mode1Up.boundsIntersect({ top, bottom }, scrollRegion);
+      const cur = {page, top, bottom, intersects};
+      if (intersects) {
+        // Also yield the page just before the visible page
+        if (prev && !prev.intersects) yield prev;
+        yield cur;
+      }
+      // Also yield the page just after the last visible page
+      else if (!cur.intersects && prev?.intersects) {
+        yield cur;
+        break;
+      }
+      prev = cur;
     }
+  }
 
-    for (const index of indicesToDisplay) {
-      const page = book.getPage(index);
-      const height = Math.floor(page.height / this.br.reduce);
-
-      if (!this.br.displayedIndices.includes(index)) {
-        const width = Math.floor(page.width / this.br.reduce);
-        const leftMargin = Math.floor((containerWidth - width) / 2);
-
-        const pageContainer = this.br._createPageContainer(index, {
-          width,
-          height,
-          top: leafTop,
-          left: leftMargin,
-        });
-
-        const pageURISrcset = this.br.options.useSrcSet ? this.br._getPageURISrcset(index, this.br.reduce, 0) : [];
-        const img = $('<img />', {
-          src: page.getURI(this.br.reduce, 0),
-          srcset: pageURISrcset,
-          alt: 'Book page image'
-        });
-        pageContainer.append(img);
-
-        BRpageViewEl.appendChild(pageContainer[0]);
+  drawLeafs() {
+    const pagesToDisplay = Array.from(this.findIntersectingPages());
```

Comment:
q: why do we need to create a new array, does `this.findIntersectingPages()` not return an array?

Reply:
It's a generator ! (#es6 #future). Array.from let's us convert a generator to an array.